Set up Tailwindcss

natsid opened the pull request:
### Description
In the process of setting up Tailwindcss, we also needed to set up:
* Webpack
* PostCSS

In order to use Webpack, created a Webpack playrunhook and added it to our build step.

Additional Tailwindcss support added:
* Some extra configuration and styles in `styles.css` to show rough examples of how to start engaging Tailwindcss
* @tailwindcss/forms plugin (an [officially supported plugin](https://tailwindcss.com/docs/plugins#official-plugins)), since UAT will obviously need good form support

### Screenshot of `/programs/new`

![Screen Shot 2021-02-23 at 5 10 03 PM](https://user-images.githubusercontent.com/5732310/108930044-06b91a00-75fa-11eb-82e7-ac485a6d5cd3.png)

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #228

natsid commented:
One question I still have is: 

Do we actually need "style-loaders" and "css-loaders" (see webpack.config.js and package.json) to get what we want?

I tried removing them and the pages appeared the same from what I could tell. But I was hard-pressed to find an example use of postcss *without* those. I'll manually test this again later with a closer eye, and if it really seems to work fine, I'll remove them.

ndmckinley commented on universal-application-tool-0.0.1/project/Webpack.scala:
```diff
@@ -0,0 +1,31 @@
+import play.sbt.PlayRunHook
```
question: Does this have to be configured in scala?  I don't really know scala.

bion replied:
Might be a good question for James Ward

natsid replied:
I honestly have no idea. I got this entire setup from the answer on [this stackoverflow post](https://stackoverflow.com/q/34568008) (and [this github repo](https://github.com/nouhoum/play-react-webpack) linked from there).

Definitely down to eliminate this lone `.scala` file if we can though.

natsid replied:
@bion Do you think I can add James as a reviewer to this PR? What's the best way to connect with them?

natsid replied:
According to James, since we're changing the sbt build, it must be in Scala.

Of course, if we move forward with #235 we could get rid of this altogether.

bion replied:
Thanks for getting in touch!

ndmckinley commented on universal-application-tool-0.0.1/webpack.sbt:
```diff
@@ -0,0 +1,17 @@
+}
+
+webpack := {
+  if(runWebpack(baseDirectory.value) != 0) throw new Exception("Something goes wrong when running webpack.")
```
nit: s/goes/went

natsid replied:
done

ndmckinley commented:
Looks like you'll need to add webpack to the container - looks like ./node_modules doesn't contain it at runtime here.

bion commented on .gitignore:
```diff
@@ -8,3 +8,4 @@ RUNNING_PID
 .bsp/
 .DS_Store
 node_modules
+bundle.js
```
question (non-blocking): thoughts on building the JS bundle somewhere other than the root directory? e.g. `build/`?

natsid replied:
It's not actually being built into the root dir. It's being built into `public/javascripts/bundle.js` (see the `path` field of `webpack.config.js`). But for whatever reason, putting the full path here ("public/javascripts/bundle.js) didn't successfully ignore the file. I can try a couple other styles of syntax, but just "bundle.js" does the trick regardless of location.

bion replied:
Ah cool, no need to change then

bion commented on universal-application-tool-0.0.1/app/views/BaseHtmlLayout.java:
```diff
@@ -29,10 +31,16 @@ public Content htmlContent(DomContent... domContents) {
-  protected Tag getCommonCssTag() {
-    return viewUtils.makeLocalCssTag("common");
+  protected Tag tailwindStyles() {
+    return viewUtils.makeLocalJsTag(WEBPACK_OUTPUT_FILENAME);
```
Question (non-blocking): seems slightly odd to load the styles via a javascript tag. Does the JS just attach a very large `<style>` tag to the DOM?

natsid replied:
I completely agree!! I tried to find out a way to avoid this, but I don't know how to compile the Tailwindcss syntax without importing it into a `.js` file... 😕 Very open to suggestions.

natsid replied:
To answer your question of whether it attaches a very large `<style>` tag: Probably? But it's hard to tell exactly what's going on. This is what the generated `bundle.js` looks like: https://gist.github.com/natsid/b33435e43008cca08e6c2868d7bc8f07.

bion replied:
Looked at this together

natsid replied:
For the record, our offline chat resulted in #235.

bion commented on universal-application-tool-0.0.1/webpack.config.js:
```diff
@@ -0,0 +1,32 @@
+                  tailwindcss
+                ],
+              },
+            },
```
Suggestion (non-blocking): extract the postcss loader to a `const`

natsid replied:
done

bion commented:
Ah cool, feel free to ignore then

On Wed, Feb 24, 2021 at 11:26 AM Natalie Sidhom <notifications@github.com>
wrote:

> *@natsid* commented on this pull request.
> ------------------------------
>
> In .gitignore
> <https://github.com/seattle-uat/universal-application-tool/pull/214#discussion_r582238949>
> :
>
> > @@ -8,3 +8,4 @@ RUNNING_PID
>  .bsp/
>  .DS_Store
>  node_modules
> +bundle.js
>
> It's not actually being built into the root dir. It's being built into
> public/javascripts/bundle.js (see the path field of webpack.config.js).
> But for whatever reason, putting the full path here
> ("public/javascripts/bundle.js) didn't successfully ignore the file. I can
> try a couple other styles of syntax, but just "bundle.js" does the trick
> regardless of location.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/seattle-uat/universal-application-tool/pull/214#discussion_r582238949>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADTURZHWNMFT2PP24LQJTLTAVHERANCNFSM4YBPOG2A>
> .
>


-- 
[image: Google Logo]
Bion Johnson (he/him)
Senior Software Engineer & Google.org Software Engineering Lead
bionj@google.com

natsid commented:
> Looks like you'll need to add webpack to the container - looks like ./node_modules doesn't contain it at runtime here.

I think I did the thing! I switched the Dockerfile to install `npm` (which has a dependency on `nodejs`) instead of just `nodejs`. And added an `npm install` command after `sbt dist`. However the last CI test run timed out... re-running.

ndmckinley commented:
niiiiiiiiice.  that works great.

ndmckinley commented:
Your lockfile is version 2 now - I think you still want that at v1.  You've run your local `npm` instead of the container's `npm`.  Maybe we should add `bin/npm` which is the same as `bin/sbt` but runs npm?

ndmckinley commented:
I notice it doesn't break the tests, but I'm suspicious (without proof) that it might be a problem later. I don't know that lockfiles are meant to be backwards incompatible indefinitely, and the little warning message that complains that it's "doing its best" doesn't inspire a ton of confidence with me...

natsid commented:
We discussed offline how to *not* run `npm` locally and added `bin/npm` so that any npm commands can be run in isolation based on the container's version of npm. And also made sure all the other scripts install npm dependencies themselves.
